### DISCUSSION TITLE
How to index composite type fields

### DISCUSSION THREAD

**Discussion by betterthanever2** [pgroonga/pgroonga#385] [Q&A] [answered]
I'm interested in creating index on Postgres composite types. 
As far as I understand, normally the process is like this:
1. create composite type & add some column of that type to some table
2. create comparison function(s)
3. create custom operators
4. create operator class that brings 2 & 3 together. 
5. create index...

I didn't find anything about this in the docs. What are the caveats, nuances? For example, following code doesn't work for me:
```
CREATE OPERATOR CLASS triple_lang_equality
    DEFAULT FOR TYPE triple_lang USING pgroonga AS
    OPERATOR 1 =(triple_lang, triple_lang),
    FUNCTION 1 triple_lang_equal(triple_lang, triple_lang);
```
Here, `triple_lang` is the name of the composite type, and `triple_lang_equal` function returns a boolean:
```
CREATE OR REPLACE FUNCTION triple_lang_equal(
    lft triple_lang,
    rght triple_lang
) RETURNS BOOLEAN AS $$
    if lft.en == rght.en and lft.ru == rght.ru and lft.uk == rght.uk:
        return True
    else:
        return False
$$ LANGUAGE plpython3u;
```

With this, I'm getting `[42P17] ERROR: invalid function number 1, must be between 1 and 0` 

What am I doing wrong?

**Comment by kou** [comment]
PGroonga doesn't support custom types.
PostgreSQL's built-in indexes such as B-Tree index supports it https://www.postgresql.org/docs/16/xindex.html .
If you want to do it, please use them instead of PGroonga.

BTW, why do you want to do it with PGroonga? It seems that `triple_lang` doesn't have any full text search target contents.

**Comment by betterthanever2** [comment]
> PGroonga doesn't support custom types.

Sounds like it's not even on the roadmap :( Sorry to hear that.

> BTW, why do you want to do it with PGroonga? It seems that triple_lang doesn't have any full text search target contents.

How do you mean? It has 3 text-type keys, each of which can hold pretty large amounts of text. Seems like a perfect full-text-search candidate to me.

**Reply by kou** [reply]
Ah, `en`/`ru`/`uk` have large text?

You can use expression index https://www.postgresql.org/docs/current/sql-createindex.html for it:

```sql
CREATE INDEX YOUR_INDEX
  ON YOUR_TABLE
  ((ARRAY[triple_lang.en, triple_lang.ru, triple_lang.uk]))
  USING pgroonga;
```

You can search it with `&@~` https://pgroonga.github.io/reference/operators/query-v2.html :

```sql
SELECT *
  FROM YOUR_TABLE
  WHERE ARRAY[triple_lang.en, triple_lang.ru, triple_lang.uk] &@~
    ('YOUR_QUERY', NULL, 'YOUR_INDEX')::pgroonga_full_text_search_condition;
```

**Comment by betterthanever2** [comment]
> You can use expression index https://www.postgresql.org/docs/current/sql-createindex.html for it

After some trial and error, here's the correct query to create index:

```
create index pgroonga_sometable_composite_type_field
    on public.sometable using pgroonga ((ARRAY[(fullname).en, (fullname).ru, (fullname).uk]));
```
The expression goes after `using`, and name of the composite field (`fullname` in this case) should be in parentheses, too.

Can you point me to the documentation on this syntax `('YOUR_QUERY', NULL, 'YOUR_INDEX')::pgroonga_full_text_search_condition` ?

**Reply by askdkc** [reply, marked as answer]
@betterthanever2 

> Can you point me to the documentation on this syntax ('YOUR_QUERY', NULL, 'YOUR_INDEX')::pgroonga_full_text_search_condition ?

You can find it here: https://pgroonga.github.io/reference/operators/query-v2.html
